### DISCUSSION TITLE
Explanation of the color indexing process? I'm getting results that are very strange to my eye.

### DISCUSSION THREAD

**Discussion by wallyfoo** [ImageMagick/ImageMagick#4022] [Digital Image Processing] [answered]
I have a tool that takes user images and indexes them to match a line of paint our company produces so our customers can pick the right paint colors for model painting. We end up using a custom index to facilitate this conversion.

One of our users uploaded an image that has a pretty solid swatch of purple, and our index swatch contains several purples that come much closer than the final result. Imagemagick chose a pale blue. Here's a screenshot of the interface and the swatch for indexing.

<img width="998" alt="Screen Shot 2021-08-02 at 4 54 49 PM" src="https://user-images.githubusercontent.com/6932210/127890565-71113ea1-db3e-4fb9-9261-a9675a7965b7.png">

![swatch_msp](https://user-images.githubusercontent.com/6932210/127890585-1cd9f098-9d55-4543-bc84-56e44f280d6f.png)

In the screenshot, you can even see that IM decided some of the dithered spaces were purple, just not the huge swath.

Since machines do what they're told, and color spaces are tricky math, I'm curious if anyone can shed some light on the choices made within imagemagick that would prioritize the blue in the result rather than several purples in the swatch that seem to be more obvious candidates?

**Comment by fmw42** [comment]
Please show your ImageMagick commands so that we know what you are doing to match the color.

**Reply by fmw42** [reply]
Also please let us know what version of ImageMagick and on what platform/OS?

**Reply by fmw42** [reply]
Also post the original image without markings and the swatch that you want to get the color. I presume that you are trying to match your swatch to one of the colors in the image of numerous different color blocks.

**Reply by wallyfoo** [reply]
That's a great idea. Here are the images. The swatch is the one above. The original and converted images are attached here.

![neon-genesis-evangelion-eva-unit-02-eva-unit-03-eva-unit-00-wallpaper-preview](https://user-images.githubusercontent.com/6932210/127895660-7b780331-2db6-48e3-bec3-034cf3124c99.jpg)

![1627917443764QTFCDc_converted](https://user-images.githubusercontent.com/6932210/127895675-8e4e6ce0-4f82-4f4f-9882-faf5cb12dade.png)

You can get these results using the following command on the desktop:
> convert neon-genesis-evangelion-eva-unit-02-eva-unit-03-eva-unit-00-wallpaper-preview.jpg -blur 2x4 -sharpen 0x3 -scale "64x64>" -dither none -map swatch_msp.png -interpolate Integer -filter point -resize "1000x1000<" -dither none test.png

Desktop: MacOS 11.4 ImageMagick 7.0.10-22 Q16 x86_64 2020-06-27

Server: Ubuntu 20.04.2 LTS ImageMagick 6.9.10-23 Q16 x86_64 20190101

The server commands are slightly different, but they get the same results:
> {uploaded_file_path}[0] -auto-orient -resize 1000x1000> -blur 2x4 -sharpen 0x3 -scale 128x128> -dither none -map {index_map_path} -interpolate Integer -filter point -resize 1000x1000< -dither none {target_path}_converted.png

**Comment by snibgo** [comment]
> -map swatch_msp.png

What does that option do? Is it a typo for "-remap swatch_msp.png"?

**Reply by wallyfoo** [reply]
Not a typo -- it works with a path to a swatch for the index colors. Perhaps it's a carry-over from an older version that aliases behind the scenes to -remap? If you run the commands above with either -map or -remap, the result is the same.

**Reply by fmw42** [reply]
Your purple swatch has a cross hair and black square on it. Should that be there in a real example? If not, please post one without. 

Why are you doing all that processing apart from the -remap -- namely all the resizing and blurring and sharpening?

What happens if you just get the average color of the swatch and then match that to your colormap colors?

**Reply by fmw42** [reply]
@snibgo -map is deprecated in favor of -remap. I would suggest they discontinue using -map. See https://imagemagick.org/script/porting.php#cli  It will work for now, but at some point will likely be removed.

**Comment by fmw42** [comment]
I took your large image and cropped an area about where you did. I then ran the following commands and get a result that to me indicated that there is a problem with IM color. Both input files are RGB as indicated by IM identify.

Swatch to test:

![purple_swatch](https://user-images.githubusercontent.com/13458218/127920160-066faf00-ee73-42bb-af03-876af5653c6b.png)



`convert purple_swatch.png -scale 1x1! -dither none -remap 127890585-1cd9f098-9d55-4543-bc84-56e44f280d6f.png -scale 200x200 swatch_mapped.png`

Which came back with this color:

![swatch_mapped](https://user-images.githubusercontent.com/13458218/127920191-c665aa6b-7da8-4a1b-adb1-848d30620480.png)


Then I tested the colors:

`convert purple_swatch.png swatch_mapped.png -scale 1x1! -format "%[pixel:u.p]\n" info:`
srgb(47.0878%,31.0079%,58.1003%)
srgb(47.0878%,31.0079%,58.1003%)

Adding -depth 8 to the above returns:
srgb(120,79,148)
srgb(120,79,148)

But measuring the color on the images from other tools says:
purple:  125,82,153
blue: 83,123,149


The IM values are exactly the same!???  Why? When they look so different.

So unless I am missing something, like the JPG was actually CMY and miss-labeled, I think there is some issue with ImageMagick.

Perhaps one of the developers or someone else might know a way to check these input images.

**Reply by fmw42** [reply]
Not an issue with color values being wrong as my mistake was pointed out by snibgo, below. So ignore the above.

**Comment by snibgo** [comment]
The choice of which colour to use seems to be in quantize.c function ClosestColor(), and is the usual Euclidean distance within the colour cube defined by the three channels:
```
distance = dR*dR + dG*dG + dB*dB
```
...where dR is the difference in the red channel, etc. It doesn't need to take the square root because we care only _which_ is the closest colour.

This means that a blue and a purple of similar intensity can be closer to each other than two purples that have different intensities.

@wallyfoo: Put this another way: if you think that one of the swatch colours is closer to that purple, which one do you think is closest? We can then manually do the distance calculations.

@fmw42 wrote:
```
convert purple_swatch.png swatch_mapped.png -scale 1x1! -format "%[pixel:u.p]\n" info:
```
For both images, this writes the value of u.p, and the "u" means use the value from the _first_ image. For useful values, remove "u.":
```
convert purple_swatch.png swatch_mapped.png -scale 1x1! -format "%[pixel:p]\n" info:

srgb(120,79,148)
srgb(83,123,149)
```

**Reply by fmw42** [reply]
Thanks, snibgo.  DOH!

**Comment by snibgo** [comment, marked as answer]
Expanding on my post above: "-remap" works on the distance within the colour cube defined by three dimensions, which by default are the sRGB channels. Instead, we can use some other dimensions such as LAB or LUV. (We can't use a colorspace with a Hue channel because that would give a false large distance between 1 degree and 359 degrees.) LAB and LUV are more perceptually uniform than sRGB, so an observer is more likely to agree the colours are close.

Windows BAT script.
```
set ColSp=LUV

%IMG7%magick ^
  127890585-1cd9f098-9d55-4543-bc84-56e44f280d6f.png ^
  -colorspace %ColSp% -set colorspace sRGB ^
  +write mpr:MAP +delete ^
  127895660-7b780331-2db6-48e3-bec3-034cf3124c99.jpg ^
  -colorspace %ColSp% -set colorspace sRGB ^
  +dither -remap mpr:MAP ^
  -set colorspace %ColSp% -colorspace sRGB ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/127934174-9a8395ef-5428-44c9-972d-3800aa6932e7.png)

**Reply by wallyfoo** [reply]
Wow! I came here merely hoping to understand what's going on under the hood, and I left with a perfect solution that completely addresses my problems. I was able to take your solution and adapt it to include my pixellation conversion. Works like a charm, _and_ I learned something in the process.

Thanks so much!
